buildkitd/buildkitd.go

```diff
@@ -135,6 +136,19 @@ func ResetCache(ctx context.Context, console conslogging.ConsoleLogger, image, c
 // MaybeStart ensures that the buildkitd daemon is started. It returns the URL
 // that can be used to connect to it.
 func MaybeStart(ctx context.Context, console conslogging.ConsoleLogger, image, containerName string, fe containerutil.ContainerFrontend, settings Settings, opts ...client.ClientOpt) (*client.Info, *client.WorkerInfo, error) {
+	if settings.StartUpLockPath != "" {
+		startLock := flock.New(settings.StartUpLockPath)
+		timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+		defer cancel()
+		_, err := startLock.TryLockContext(timeoutCtx, 200*time.Millisecond)
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return nil, nil, errors.Errorf("timeout waiting for other process to start buildkitd")
+			}
+			return nil, nil, errors.Wrapf(err, "try flock context %s", settings.StartUpLockPath)
+		}
+		defer startLock.Unlock()
+	}
```

Comment:
FWIW, I tried first using a shared lock, which would later be upgraded to an exclusive lock, but it seems that on some systems the shared lock becomes automatically an exclusive lock, which causes strange behaviors.

The logic I ended up using (exclusive locking only) seems to be performant enough in practice (there's no bottleneck that I could notice in my experimentation locally).